src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+
+  DeviceHandle = hdInitDevice(this->DeviceName.c_str());
+
+  HDErrorInfo error;
+  if (HD_DEVICE_ERROR(error = hdGetError()))
+  {
+    LOG_ERROR( "Failed to initialize Phantom Omni " << this->DeviceName);
+    DeviceHandle = -1;
+    return PLUS_FAIL;
+  }
+
+  // Enable forces
+  hdEnable(HD_FORCE_OUTPUT);
+  hdEnable(HD_FORCE_RAMPING);
+
+  this->isHapticDeviceInitialized = true;
+  LOG_INFO("Phantom initialized: " << this->DeviceName)
+
+  hdStartScheduler();
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalDisconnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    hdDisableDevice(DeviceHandle);
+    hdStopScheduler();
+    this->isHapticDeviceInitialized=false;
+  }
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::ReadConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, rootConfigElement);
+  XML_READ_CSTRING_ATTRIBUTE_REQUIRED(DeviceName,deviceConfig);
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(trackerConfig, rootConfigElement);
+  trackerConfig->SetAttribute("DeviceName", this->DeviceName.c_str());
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStartRecording()
+{
+  if (!this->isHapticDeviceInitialized)
+  {
+    LOG_ERROR("InternalStartRecording failed: OpenHaptics device has not been initialized");
+    return PLUS_FAIL;
+  }
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStartRecording");
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStopRecording()
+{
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStopRecording");
+  return PLUS_SUCCESS;
+}
+
+HDCallbackCode HDCALLBACK
+vtkPlusOpenHapticsDevice::positionCallback(void* pData)
+{
+  
+  vtkPlusOpenHapticsDevice* client = reinterpret_cast<vtkPlusOpenHapticsDevice*>(pData);
+  HHD handle = client->DeviceHandle;
+  HD_state state = client->DeviceState;
+
+  const double unfilteredTimestamp = vtkPlusAccurateTimer::GetSystemTime();
+  ++client->FrameNumber;
+
+  HDdouble force[3];
+  force[0] = 0;
+  force[1] = 0;
+  force[2] = 0;
+
+  //Transform to RAS coordinate system
+  vtkSmartPointer<vtkMatrix4x4> rasCorrection = vtkSmartPointer<vtkMatrix4x4>::New();
+  rasCorrection->Identity();
+  rasCorrection->SetElement(0, 0, -1);
+  rasCorrection->SetElement(1, 1, 0);
+  rasCorrection->SetElement(1, 2, 1);
+  rasCorrection->SetElement(2, 2, 0);
+  rasCorrection->SetElement(2, 1, 1);
+
+  //assemble force data
+  auto inputChannels = client->InputChannels;
+  vtkPlusDataSource* forceInput;
+  if (!inputChannels.empty() && inputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))
+  {
+    StreamBufferItem* item = new StreamBufferItem();
+    if (forceInput->GetLatestItemUidInBuffer() > 0)
+    {
+      forceInput->GetLatestStreamBufferItem(item);

```

Comment:
Check that the return value is ITEM_OK, and also that the transform is valid with item->GetStatus()